### PR TITLE
CNCF cloudevent support for skaffold events

### PR DESCRIPTION
**Pull request by baleshk**
Fixes: #4259

**Description**
In this pull request, we have added support for cloudevents standards on a different route /v3/events.

### CODE REVIEW

**Comment by aaron-prindle** [comment]
To help reviewers, it might make sense to outline what the PR does more directly to help understand which files are most important in the list of 81 files.  Would it be possible to add some more context for revieweres,  perhaps something like what I have started below?:

- Added eventv3.<Event> function calls everywhere eventv2.<Event> function calls were (plumbed v3 events throughout code where v2 events were)
- Added `CheckDeepEqualProtoMessage` checks and replaced old checks where relevant s/CheckDeepEqual/CheckDeepEqualProtoMessage
- Updated hack/generate-proto.sh to support generated `eventv3` protos
- NOTE: some eventv2 boilerplate is exactly copied over for `eventv3`, some changes are novel.  The important files with modifications/important-differences include:
  - X.go, Y.go, z.proto

Critical files:
- X.go
- Y.go
- Z.go

**Comment by aaron-prindle** [comment]
.DS_STORE file should be removed

File: pkg/skaffold/server/v3/endpoints.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	resetStateOnDeploy = event.ResetStateOnDeploy
+)
+
+func (s *Server) GetState(context.Context, *empty.Empty) (*proto.State, error) {
+	return event.GetState()
+}
+
+func (s *Server) Events(_ *empty.Empty, stream proto.SkaffoldV3Service_EventsServer) error {
+	return event.ForEachEvent(func(event *proto.Event) error {
+		event.Datacontenttype = "text/xml"
```
Comment (aaron-prindle) [inline pkg/skaffold/server/v3/endpoints.go:43]: Can you add a comment explaining why `text/xml` is used?  I remember there was some conversation around this during the doc presentation, @briandealwis did you have anything to add here?

File: proto/v2/markdown.tmpl [hunk 2]
```diff
@@ -80,4 +80,4 @@ You can find the source for {{.Name}} [on Github](https://github.com/GoogleConta
 {{end}} <!-- end HasExtensions -->
 
 
-{{end}}
+{{end}}
```
Comment (aaron-prindle) [inline proto/v2/markdown.tmpl:83]: nit: is removing the newline required?  Why this change?

File: pkg/skaffold/server/server.go [hunk 3]
```diff
@@ -196,9 +207,21 @@ func newGRPCServer(preferredPort int, usedPorts *util.PortSet) (func() error, in
 }
 
 func newHTTPServer(preferredPort, proxyPort int, usedPorts *util.PortSet) (func() error, error) {
-	mux := runtime.NewServeMux(runtime.WithProtoErrorHandler(errorHandler), runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{OrigName: true, EmitDefaults: true}))
```
Comment (aaron-prindle) [inline pkg/skaffold/server/server.go:199]: What is the reason the errorHandler is removed?

File: pkg/skaffold/server/server.go [hunk 4]
```diff
@@ -231,21 +259,6 @@ func newHTTPServer(preferredPort, proxyPort int, usedPorts *util.PortSet) (func(
 	}, nil
 }
 
-type errResponse struct {
-	Err string `json:"error,omitempty"`
-}
-
-func errorHandler(ctx context.Context, _ *runtime.ServeMux, marshaler runtime.Marshaler, writer http.ResponseWriter, _ *http.Request, err error) {
```
Comment (aaron-prindle) [inline pkg/skaffold/server/server.go:238]: What is the reason the errorHandler is removed?

**Comment by aaron-prindle** [comment]
Also currently this PR needs to be rebased to resolve some conflicts with file at `main` branch.  Lmk if you have any question on how to resolve merge issues

**Comment by aaron-prindle** [comment]
Also can you re-push your PR?   For some reason our Github Actions aren't running against it

File: pkg/skaffold/server/server.go (hunk 3 above)
Comment (baleshk) [reply pkg/skaffold/server/server.go:199]: In grpc-gateway v2, Error handling configuration has been overhauled.
_runtime.HTTPError, runtime.OtherErrorHandler, runtime.GlobalHTTPErrorHandler, runtime.WithProtoErrorHandler are all gone. Error handling is rewritten around the use of gRPCs Status types._
Custom Error handling function may not be required anymore.

File: pkg/skaffold/server/server.go (hunk 4 above)
Comment (baleshk) [reply pkg/skaffold/server/server.go:238]: Same as above, In grpc-gateway v2, Error handling configuration has been overhauled.
_runtime.HTTPError, runtime.OtherErrorHandler, runtime.GlobalHTTPErrorHandler, runtime.WithProtoErrorHandler are all gone. Error handling is rewritten around the use of gRPCs Status types._
Custom Error handling function may not be required anymore.

File: proto/v2/markdown.tmpl (hunk 2 above)
Comment (baleshk) [reply proto/v2/markdown.tmpl:83]: No newline is removed. Dont know why it captured (removal and addition of newline). Will fix it.

File: pkg/skaffold/server/v3/endpoints.go (hunk 1 above)
Comment (baleshk) [reply pkg/skaffold/server/v3/endpoints.go:43]: Fixed. It should be _application/json_. 
Following comment added- 
_grpc gateway converts messages from protobuf format to JSON format, so updating DatacontentType to application/json_

File: pkg/skaffold/server/server.go (hunk 3 above)
Comment (briandealwis) [reply pkg/skaffold/server/server.go:199]: I wonder how we can test this

**Comment by briandealwis** [comment]
To help break this up, I’m taking on separating out the grpc-gateway upgrade. This update surfaced hundreds (thousands?) of warnings as we are passing protobuf objects by-value which is a bad practice. I’ll address those and then @baleshk will rebase/meter in that change.

**Comment by baleshk** [comment]
> .DS_STORE file should be removed

Will take care in the new pull request.

File: pkg/skaffold/server/v3/endpoints.go (hunk 1 above)
Comment (baleshk) [reply pkg/skaffold/server/v3/endpoints.go]: This issue is fixed now.

**Comment by aaron-prindle** [comment]
This PR depends on the changes in:
https://github.com/GoogleContainerTools/skaffold/pull/6567

**Comment by aaron-prindle** [comment]
@baleshk can you rebase/fix-the-conflicts in this PR with:
pkg/skaffold/kubernetes/status/resource/deployment.go

Additionally it might make sense to rebase off of @briandealwis PR here - https://github.com/GoogleContainerTools/skaffold/pull/6567

if there are any changes from how you updated `grpc-gateway` to v2

File: proto/v3/skaffold.proto [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+// It is a wrapper on the top of other events like MetaEvent, BuildEvent, TestEvent, DeployEvent, PortEvent, StatusCheckEvent, ResourceStatusCheckEvent, FileSyncEvent, or DebuggingContainerEvent.
+message Event {
+    string specversion = 1;
+    string type = 2;
+    string source = 3;
+    string id = 4;
+    google.protobuf.Timestamp time = 5;
+    google.protobuf.Any data = 6;
+    string datacontenttype = 7;
+}
```
[the remark this answers is not in the document]
Comment (baleshk) [reply proto/v3/skaffold.proto:156]: @briandealwis We are following [CloudEvents Spec](https://github.com/cloudevents/spec) and not [CloudEvents Protobuf Spec](https://github.com/cloudevents/spec/blob/v1.0.1/protobuf-format.md).
Reason for not going with CloudEvent Protobuf spec - Not all major Language's SDK support CloudEvent protobuf specs. For instance, while [Java SDK](https://cloudevents.github.io/sdk-java/protobuf.html) has support for this but not [Golang](https://github.com/cloudevents/sdk-go)

[the remark this answers is not in the document]
Comment (briandealwis) [reply proto/v3/skaffold.proto:156]: The CE Go SDK [does support protobuf format](https://github.com/cloudevents/sdk-go/blob/236e4413fbc4508b54cf5c45826adf2c4af61376/binding/format/protobuf/v2/protobuf.go) but it's undocumented.

That said, I don't know how this would integrate into grpc-framework.

**Comment by baleshk** [comment]
Please refer to https://github.com/GoogleContainerTools/skaffold/pull/6679 (which is rebased/merged with the head) and has some more changes on the top of this one.
